ebi_eva_common_pyutils/nextflow/nextflow_pipeline.py

```diff
@@ -0,0 +1,174 @@
+            for root_process_in_dependent_pipeline in dependent_pipeline._get_root_processes():
+                joined_pipeline.add_process_dependency(root_process_in_dependent_pipeline,
+                                                       final_process_in_main_pipeline)
+        return joined_pipeline
+
+    def _get_root_processes(self) -> List[NextFlowProcess]:
+        # Root processes are those which have no dependencies
+        # See https://stackoverflow.com/a/62948641
+        roots = []
+        for component in nx.weakly_connected_components(self.process_dependency_map):
+            subgraph = self.process_dependency_map.subgraph(component)
+            roots.extend([n for n, d in subgraph.out_degree() if d == 0])
+        return roots
+
+    def _get_final_processes(self) -> List[NextFlowProcess]:
+        # Final processes are those which have no other processes depending on them
+        # See https://stackoverflow.com/a/62948641
+        roots = []
+        for component in nx.weakly_connected_components(self.process_dependency_map):
+            subgraph = self.process_dependency_map.subgraph(component)
+            roots.extend([n for n, d in subgraph.in_degree() if d == 0])
+        return roots
+
+    @staticmethod
+    def _get_process_repr(process: NextFlowProcess, dependencies: List[NextFlowProcess]) -> str:
+        process_directives_str = "\n".join([f"{key}='{value}'" for key, value in process.process_directives.items()])
+        input_dependencies = "val flag from true"
+        if dependencies:
+            input_dependencies = "\n".join([f"val {dependency.success_flag} from {dependency.success_flag}"
+                                            for dependency in dependencies])
+        return "\n".join(map(str.strip, f"""
+                    process {process.process_name} {{
+                    {process_directives_str}
+                    input:
+                    {input_dependencies}
+                    output:
+                    val true into {process.success_flag}
+                    script:
+                    \"\"\"
+                    {process.command_to_run}
+                    \"\"\"
+                    }}""".split("\n")))
+
+    def __str__(self):
+        # Order the list of nodes based on the dependency
+        # See https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.traversal.depth_first_search.dfs_postorder_nodes.html?highlight=dfs_postorder_nodes#networkx.algorithms.traversal.depth_first_search.dfs_postorder_nodes
+        ordered_list_of_processes_to_run = list(nx.dfs_postorder_nodes(self.process_dependency_map))
+        # Get a Nextflow pipeline representation of each process and its dependencies
+        return "\n\n".join([NextFlowPipeline._get_process_repr(process, list(self.process_dependency_map[process]))
+                            for process in ordered_list_of_processes_to_run])
+
+
+class LinearNextFlowPipeline(NextFlowPipeline):
```

Comment:
Linear pipeline is now represented as a simplified case of the general Nextflow pipeline.